tests/functional/olp-cpp-sdk-authentication/DecisionApiTest.cpp

```diff
@@ -0,0 +1,209 @@
+      unsigned int expires_in = kLimitExpiry, bool do_cancel = false) {
+    std::shared_ptr<AuthenticationClient::SignInClientResponse> response;
+    unsigned int retry = 0u;
+    do {
+      if (retry > 0u) {
+        OLP_SDK_LOG_WARNING(kLogTag,
+                            "Request retry attempted (" << retry << ")");
+        std::this_thread::sleep_for(
+            std::chrono::seconds(retry * kRetryDelayInSecs));
+      }
+
+      std::promise<AuthenticationClient::SignInClientResponse> request;
+      auto request_future = request.get_future();
+
+      AuthenticationClient::SignInProperties props;
+      props.expires_in = std::chrono::seconds(expires_in);
+
+      auto cancel_token = client_->SignInClient(
+          credentials, props,
+          [&](const AuthenticationClient::SignInClientResponse& resp) {
+            request.set_value(resp);
+          });
+
+      if (do_cancel) {
+        cancel_token.Cancel();
+      }
+      response = std::make_shared<AuthenticationClient::SignInClientResponse>(
+          request_future.get());
+    } while ((!response->IsSuccessful()) && (++retry < kMaxRetryCount) &&
+             !do_cancel);
+
+    return *response;
+  }
+
+  AuthorizeResponse Authorize(const std::string& access_token,
+                              AuthorizeRequest request,
+                              bool do_cancel = false) {
+    std::shared_ptr<AuthorizeResponse> response;
+    auto retry = 0u;
+    do {
+      if (retry > 0u) {
+        OLP_SDK_LOG_WARNING(kLogTag,
+                            "Request retry attempted (" << retry << ")");
+        std::this_thread::sleep_for(
+            std::chrono::seconds(retry * kRetryDelayInSecs));
+      }
+
+      std::promise<AuthorizeResponse> resp;
+      auto request_future = resp.get_future();
+
+      auto cancel_token = client_->Authorize(
+          access_token, std::move(request),
+          [&](const AuthorizeResponse& responce) { resp.set_value(responce); });
+
+      if (do_cancel) {
+        cancel_token.Cancel();
+      }
+      response = std::make_shared<AuthorizeResponse>(request_future.get());
+    } while ((!response->IsSuccessful()) && (++retry < kMaxRetryCount) &&
+             !do_cancel);
+
+    return *response;
+  }
+
+  std::string GetErrorId(
+      const AuthenticationClient::SignInUserResponse& response) const {
+    return response.GetResult().GetErrorResponse().error_id;
+  }
+};
+
+TEST_F(AuthenticationClientTestAuthorize, AuthorizeAllow) {
+  AuthenticationCredentials credentials(id_, secret_);
+  auto resp = SignInClient(credentials, kExpiryTime);
+  auto res = resp.GetResult();
+  EXPECT_TRUE(resp.IsSuccessful());
+  EXPECT_EQ(olp::http::HttpStatusCode::OK, res.GetStatus());
+
+  auto token = res.GetAccessToken();
+  auto request = olp::authentication::AuthorizeRequest().WithServiceId(service_id_);
+  request.WithAction("getTileCore");
+  auto response = Authorize(token, std::move(request));
+  auto result = response.GetResult();
```

Comment:
No need to copy Result
```suggestion
  const auto& result = response.GetResult();
```